Form validation: Reduce aggressiveness of onfocusout validation

EricDunsworth opened the pull request:
This partially-reverts #7921.

One of its changes caused form validation to always get triggered when focusing out of fields. The end result was that required field validation became drastically more aggressive, arguably to the point of impeding usability.

For example:
* Tabbing through required inputs, checkboxes or radio buttons that had never received any user input would immediately trigger required field errors
* Ticking and then unticking any required checkbox would immediately trigger a required field error

Examples of situations where that would annoy users:
* Keyboard user getting bombarded with required field errors while tabbing through a form
* Mouse/Touch screen user accidentally clicking or touching a random required field, then having to deal with fields visually-shifting downwards

According to the jQuery Validation Plugin's reference documentation (https://jqueryvalidation.org/reference/#link-validation-event), the plugin is meant to provide quick contextual error feedback, without going so far as to annoy users. It considers displaying error messages without having received user input to cross that line.

I personally find the behaviour I'm reverting to be really annoying and aren't the only one that dislikes it... That behaviour was also brought up as a concern by a usability subject matter expert that reviewed a form project I'm currently helping out with.

Fixes #8353 and partially-addresses #8688.

Related to #8111, #8160 and #8066.

CC @biljanap @RobJohnston @duboisp @thekodester @sproules @juleskuehn @lisafast @NoisyMonk @marieeved @rlambert27 @andrewnordlund @shawnthompson

RobJohnston commented:
Makes sense to me.  Also related to #7394.

biljanap commented:
I am looking forward for this being resolved. It's still annoying me and everybody else involved. It's constantly being flagged as a bug.

EricDunsworth committed:
Form validation: Reduce aggressiveness of onfocusout validation

This partially-reverts #7921.

One of its changes caused form validation to always get triggered when focusing out of fields. The end result was that required field validation became drastically more aggressive, arguably to the point of impeding usability.

For example:
* Tabbing through required inputs, checkboxes or radio buttons that had never received any user input would immediately trigger required field errors
* Ticking and then unticking any required checkbox would immediately trigger a required field error

Examples of situations where that would annoy users:
* Keyboard user getting bombarded with required field errors while tabbing through a form
* Mouse/Touch screen user accidentally clicking or touching a random required field, then having to deal with fields visually-shifting downwards

According to the jQuery Validation Plugin's reference documentation (https://jqueryvalidation.org/reference/#link-validation-event), the plugin is meant to provide quick contextual error feedback, without going so far as to annoy users. It considers displaying error messages without having received user input to cross that line.

I personally find the behaviour I'm reverting to be really annoying and aren't the only one that dislikes it... That behaviour was also brought up as a concern by a usability subject matter expert that reviewed a form project I'm currently helping out with.

Fixes #8353 and partially-addresses #8688.

Related to #8111, #8160 and #8066.

lisafast commented:
Agree - very confusing and annoying for users to get errors before finishing the form.

EricDunsworth commented:
Btw I just wanted to mention that even with this PR in place, there will still be some circumstances in which fields will be validated before a user attempts to submit a form. But those will only be triggered when a user really has messed up - like if they've inserted an invalid value.

For example:
* If a user types a letter into a number field, then focuses away, an error will immediately appear
* If a field is already highlighted as having an error, the field will be re-validated in realtime each time anything is typed into it


Here's an excerpt of the [part I linked to in the OP](https://jqueryvalidation.org/reference/#link-validation-event) that covers how the plugin is intended to work:
> **Validation event**
> 
> By default, forms are validated on submit, triggered by the user clicking the submit button or pressing enter when a form input is focused (option ``onsubmit``). In addition, once a field was highlighted as being invalid, it is validated whenever the user types something in the field (option ``onkeyup``). When the user enters something invalid into a valid field, it is also validated when the field loses focus (option ``onfocusout``).
> 
> The goal of these interactions is to provide feedback as early as possible, whilst avoiding user annoyance. Displaying error messages before the user had the chance to even type something is not helpful.

EricDunsworth commented:
Just went over this with @delisma and @GormFrank. They agree that this PR should be the form validation plugin's default behaviour in WET. But they want to add a configuration option to optionally make it aggressive (i.e. #7921's behaviour).

So I'll either revise this PR to do that or open a separate PR to add the extra option.

GormFrank commented:
@bsouster ,
@duboisp would like to discuss with you on this topic.

EricDunsworth commented:
I tried looking into adding an extra configuration option earlier, but haven't been able to figure out how to properly implement it to my liking. jQuery validation has [its own built-in ``onfocusout`` method](https://jqueryvalidation.org/validate/#onfocusout) that only kicks in in specific scenarios. Not sure how to go about modifying it to retain its standard behaviour by default and replace it with aggressive behaviour if the extra config option is enabled.

I'm personally against adding this config option since IMO its sole purpose would be to annoy users and impede usability. Pretty much everyone agrees that the status quo isn't ideal, Plus jQuery Validation was never intended to validate this aggressively and its documentation even advocates against what WET is currently doing.

IMO the config option shouldn't be a show-stopper to merge in this PR. I believe the usability benefits on all (or the vast majority of) WET forms would far outweigh the need to add an extra configuration option that activates undesired behaviour.

@GormFrank @delisma Would one of you be willing to add the custom configuration option if you strongly believe in having it available?

duboisp commented:
We need to evaluate the accessibility impact (evidence) and state formally what is the best default behaviour.

Initially that feature was inserted to improve accessibility for screen reader and it was supported by @neoinsight in 2017. But seeing this to be a reverse opinion now, that means we are missing some evidence. This change suggestion would need to be supported by UX research or/and expertise. After those document published, we would be able to discuss what behaviour is the best for the default and what behaviour should be optional.

duboisp commented:
@lisafast can you provide the evidence of your testing.

bsouster commented:
I have no evidence to have a preference for one or the other. I do think the validation error should wait until you leave a set of options (checkbox or radio) to trigger instead of after tabbing out of the first option element of a set of options

GormFrank commented:
Waiting on Eric's design decision on validation aggressiveness for this one.

EricDunsworth commented:
Finally got around to sending wet-boew/wet-boew-documentation#262 :).

juleskuehn commented:
> 
> 
> @GormFrank Finally got around to sending [wet-boew/wet-boew-documentation#262](https://github.com/wet-boew/wet-boew-documentation/pull/262).

I couldn't find the decision document itself? (13.html)?

EricDunsworth commented:
@juleskuehn Looks like I forgot to add it to my working tree before committing lol. It got buried among a bunch of other untracked files I had there. I'll sneak it in a moment.

EricDunsworth commented:
@juleskuehn Added :)! Btw thanks for having spotted that.

duboisp commented:
La [décision de conception DD-13](https://wet-boew.github.io/wet-boew-documentation/decision/13.html) a été présenté au dernier WET roadmap et est en bonne voie pour une approbation à la prochaine réunion. Si nécessaire on fera des ajustements ultérieurements.
